src/engine/fileOpsIns.cpp

```diff
@@ -549,89 +577,173 @@ void DivEngine::loadSBI(SafeReader& reader, std::vector<DivInstrument*>& ret, St
-      opC.vib = ((sbi_Ccharacteristics >> 6) & 0x1);
-      opC.am = ((sbi_Ccharacteristics >> 7) & 0x1);
-      opC.tl = sbi_Cscaling_output & 0x3F;
-      opC.ksl = ((sbi_Cscaling_output >> 6) & 0x3);
-      opC.ar = ((sbi_Ceg_AD >> 4) & 0xF);
-      opC.dr = (sbi_Ceg_AD & 0xF);
-      opC.rr = (sbi_Ceg_SR & 0xF);
-      opC.sl = ((sbi_Ceg_SR >> 4) & 0xF);
-      opC.ws = sbi_Cwave;
-      
-      opM4.mult = sbi_M4characteristics & 0xF;
-      opM4.ksr = ((sbi_M4characteristics >> 4) & 0x1);
-      opM4.sus = ((sbi_M4characteristics >> 5) & 0x1);
-      opM4.vib = ((sbi_M4characteristics >> 6) & 0x1);
-      opM4.am = ((sbi_M4characteristics >> 7) & 0x1);
-      opM4.tl = sbi_M4scaling_output & 0x3F;
-      opM4.ksl = ((sbi_M4scaling_output >> 6) & 0x3);
-      opM4.ar = ((sbi_M4eg_AD >> 4) & 0xF);
-      opM4.dr = (sbi_M4eg_AD & 0xF);
-      opM4.rr = (sbi_M4eg_SR & 0xF);
-      opM4.sl = ((sbi_M4eg_SR >> 4) & 0xF);
-      opM4.ws = sbi_M4wave;
-
-      opC4.mult = sbi_C4characteristics & 0xF;
-      opC4.ksr = ((sbi_C4characteristics >> 4) & 0x1);
-      opC4.sus = ((sbi_C4characteristics >> 5) & 0x1);
-      opC4.vib = ((sbi_C4characteristics >> 6) & 0x1);
-      opC4.am = ((sbi_C4characteristics >> 7) & 0x1);
-      opC4.tl = sbi_C4scaling_output & 0x3F;
-      opC4.ksl = ((sbi_C4scaling_output >> 6) & 0x3);
-      opC4.ar = ((sbi_C4eg_AD >> 4) & 0xF);
-      opC4.dr = (sbi_C4eg_AD & 0xF);
-      opC4.rr = (sbi_C4eg_SR & 0xF);
-      opC4.sl = ((sbi_C4eg_SR >> 4) & 0xF);
-      opC4.ws = sbi_C4wave;
+        readSbiOpData(sbi_op12, reader);
+
+        ins = new DivInstrument;
+        DivInstrumentFM::Operator& opM6 = ins->fm.op[0];
+        DivInstrumentFM::Operator& opC6 = ins->fm.op[1];
+        ins->type = DIV_INS_OPL;
+        ins->fm.ops = 2;
+        ins->name = fmt::format("{0} (2op)", patchName);
+        writeOp(sbi_op12, opM6, opC6);
+        ins->fm.alg = (sbi_op12.FeedConnect & 0x1);
+        ins->fm.fb = ((sbi_op12.FeedConnect >> 1) & 0x7);
+      }
 
       // Ignore rest of file once we've read in all we need.
       // Note: Freq Monster 801 adds a ton of other additional fields irrelevant to chip registers.
+      //       If instrument transpose is ever supported, we can read it in maybe?
       reader.seek(0, SEEK_END);
+      ret.push_back(ins);
     }
 
   } catch (EndOfFileException& e) {
     lastError = "premature end of file";
     logE("premature end of file!");
     delete ins;
-    return;
   }
+}
+void DivEngine::loadBNK(SafeReader& reader, std::vector<DivInstrument*>& ret, String& stripPath) {
+  DivInstrument* insList[256];
+  String instNames[256];
+  memset(insList, 0, 256 * sizeof(void*));
+  reader.seek(0, SEEK_SET);
 
-  ret.push_back(ins);
+  // First distinguish between GEMS BNK and Adlib BNK
+  uint64_t header = reader.readL();
+  bool is_adlib = ((header>>8) == 0x2d42494c444100L);
+  int readCount = 0;
+
+  if (is_adlib) {
+    // Caveat: Technically Adlib BNK can hold up to 0xFFFF instruments, 
```

Comment:
Caveat fix: Use std::vector<DivInstrument*> insList. Then we can display all 65 thousand instruments to the user when the new instrument selector comes.

Reply:
Done, tested and works!